python/rmm/rmm.py

```diff
@@ -231,3 +238,21 @@ def rmm_cupy_allocator(nbytes):
     ptr = cupy.cuda.memory.MemoryPointer(mem, 0)
 
     return ptr
+
+
+def register_reinitialize_hook(func, *args, **kwargs):
+    """
+    Register a hook to be called by `rmm.reinitialize()`.
+
+    Hooks are called in the *reverse* order they are registered.
+    """
+    _reinitialize_hooks[func] = (args, kwargs)
```

Comment:
I know that it typically works, but I'm not the biggest fan of using callables as dict keys. It's also quite easy to construct realistic hooks that won't be hashable (e.g. functors without `__hash__`). Would it be too onerous to require users to provide a name for hooks? That would also give us the ability to provide a better UX for something like listing all available hooks if we chose to add that kind of feature.

Reply:
Yeah using something ordered (perhaps `list` or `deque`) seems like a good idea as users may expect these to be run in a particular order

Reply:
Dict's are insertion-ordered as of Python 3.7, so we're already good on the ordering front (and it is promised by the registration API as seen in the docs above). I'm pro switching to a list though.

Reply:
Indeed ideally we would want something that behaves like a stack (or at least is used like one). IOW LIFO

Reply:
> What is a hook intended to be? Deregistration/clearing function for a single cudf (or other library) allocation, or something broader like cudf.clear_cache()? 

The motivation for this PR is indeed https://github.com/rapidsai/cudf/issues/11237 (`cudf.clear_cache()`)

That being said, I think the behavior should match `atexit` as closely as possible as it's the "hook" registration/unregistration behavior people are most likely to be familiar with, and thus most likely to expect. 

So:

* LIFO
* Hooks can be registered multiple times (with different args/kwargs)
* Unregistering a hook unregisters all instances of hooks
* When unregistering, function comparisons are done via `==`

Reply:
Thanks @pentschev. You're right that we can't guarantee any particular order in which `cudf.clear_cache()` and `cupy.clear_cache()` will be called, if both cuDF and CuPy internally register their hooks with RMM at import time. It would depend on the order in which they are imported. 

> we should at least document what a hook's objective really is,

How about something along these lines?

```python
def register_reinitialize_hook(func, *args, **kwargs):
    """
    Add a function to the list of functions that will be called before `rmm.reinitialize()`.

    Typically, a library will use this function to register hooks that are responsible
    for deleting any remaining internal references to objects that use RMM to
    allocate device memory.
```

Reply:
> Ideally, the two functions will not be written to depend on each other.

If they do, this is "inappropriate intimacy" in the lingo of the OO/refactoring world.

I think the idea is that when you call `reinitialize`, a new `MemoryResource` is created. If no allocations are made with this `MemoryResource` then the only reference to it is internal to RMM. This reference is dropped if we call `reinitialize` a second time.

When an allocation is made via `DeviceBuffer`, it holds a reference to the memory resource it was allocated from, so as long as the `DeviceBuffer` object is live, the memory resource is live too. Consequently, if an external library holds on to a cache of `DeviceBuffer` objects, then the memory resource will persist through `reinitialize`.

So to address @pentschev's point about deregistering allocations, I think that the hooks are responsible for dropping references to `DeviceBuffer` objects that the user got a reference to (in the cudf case, the motivating case is "interning" scalars as device memory). They (should?) never  touch the memory resource directly, and so ordering between libraries is not problematic.

Reply:
> Unless we have arguments opposed to the problem above, we should at least document what a hook's objective really is, i.e., deallocate single allocation vs deallocate all allocations from a given library.

I think this a big hammer, for dropping reference to all allocations from a given library. We are relying on the usual garbage collection/refcount mechanisms for fine-grained deallocation, the problem this is trying to solve is global caches (one of the two hardest problems in computer science)

Reply:
As long as we don't provide any guarantees regarding the order hooks are called, which we seem to be leaning towards, then I agree with the arguments above. My concern was mainly regarding the hook ordering, if we guarantee the order in which they are called that implies a having dependencies between libraries is acceptable.

Reply:
> This is valuable, for example, when a library registers multiple hooks (that are independent of any other library's hooks) and needs them to run in a specific order for cleanup to be safe.

This is fair, and circles back to the scope of what I said about defining what a hook really is allowed to do, which you just did. :)

Reply:
Thanks, Peter! Seems like we are all in alignment here. Going to mark this one resolved